Integrating cagecleaner

LucoDevro opened the pull request:
Hi Cameron,

cagecleaner has been released and has been integrated into cblaster! I have included it as an additional step within the search subcommand, both in the CLI and the GUI. I have already done the necessary rewiring: cblaster should be able to call cagecleaner automatically and parse its results smoothly within its own workflow.

Major changes that were required for this integration:

- cblaster now has 8 dependencies, including cagecleaner, so I recommend using a conda environment now. I have added requests (the sole dependency of cblaster) to the yml of cagecleaner and distributed it in this repo as env.yml.
- Installation instructions in the README have been updated as well.
- Other updates of the README, including referrals to the cagecleaner wiki.

Some bugs in cblaster I fixed right away:

- The GUI command window did not prints logs on separate lines.
- Launching fresh search sessions from the GUI did not yield a session json.

I'm finalising the application note, so the citations will be updated shortly.

gamcil commented:
Hi @LucoDevro, sorry for dragging my feet on this. The integration itself looks mostly good, my main concern here is the conda environment and a few other points.

* We already have a [bioconda environment](https://github.com/bioconda/bioconda-recipes/blob/fca26079bf79b9564fcc32405dc69eaf94b3cdb4/recipes/cblaster/meta.yaml) which we should update instead of a local environment file. The cleanest way would be for you to make a Bioconda recipe for CAGEcleaner which I can then add as a dependency for cblaster.
* The dependency list in your env file also looks very long, which ones are actually necessary for CAGEcleaner?
* The bioconda recipe for skDER (and one of its dependencies PhiSpy) doesn't seem to support Apple Silicon yet. I think their recipe needs to be adjusted something [like this](https://github.com/bioconda/bioconda-recipes/pull/50207/files), so maybe we can submit a PR
* cblaster should still be installable/function if CAGEcleaner/skDER cannot be installed. We can use meta selectors in the bioconda recipe to specifically not install CAGEcleaner on osx-arm64. Within your PR, perhaps we can [treat it like genomicsqlite](https://github.com/gamcil/cblaster/blob/master/cblaster/database.py#L19) so the dereplication is always disabled if CAGEcleaner is not installed 
* Is CAGEcleaner importable as a library? If so, it would be neater to just import it and call it directly from within cblaster rather than as a subprocess
* For the temporary files, you currently make a directory in the current working directory. For this type of thing, I have previously preferred using e.g. NamedTemporaryFile from the tempfile library (see [1](https://github.com/gamcil/cblaster/blob/6ada227343b3b57cea4f793a63c119e4368ec7e0/cblaster/hmm_search.py#L255) and [2](https://github.com/gamcil/cblaster/blob/6ada227343b3b57cea4f793a63c119e4368ec7e0/cblaster/local.py#L161)). Perhaps you could use [TemporaryDirectory](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory) here, and then copy files across when `--keep_dereplication_files` is true. This would ensure the directory is unique and is automatically deleted
* Minor point, I generally prefer using Pathlib instead of os

LucoDevro commented:
Hi @gamcil, a quick update on your requested/suggested changes:

- CAGEcleaner now has a Bioconda recipe. It passed the checks at Bioconda and is now waiting for approval.
- I have cleaned the dependency list. The previous version was a raw conda env export.
- I have filed an [issue at skDER](https://github.com/raufs/skDER/issues/9) to do changes to their recipe as you suggested. These are now also awaiting approval at Bioconda. (PhiSpy is indeed a dependency of skDER but its functionality is not used by CAGEcleaner.)
- This is not relevant anymore because of the above.
- No, it is not. I designed it as a pure CLI tool for use in the terminal, and I would have to do some more efforts to make it cleanly importable in Python. If there is sufficient demand, I'll take a look into it.
- Thanks for the suggestion -> implemented.
- Why Pathlib instead of os?

Change commits and updated conda environments will follow once all Bioconda approvals have been done.
